src/components/profile/changeUserEmail.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { useState, useEffect } from 'react';
+import { AxiosError } from 'axios';
+import { useQuery } from '@tanstack/react-query';
+import ChangeEmailData from '../../types/ChangeEmailData';
+import ServerError from '../../types/ServerError';
+import { fetchUser } from '../../api/client';
+import useChangeUserEmail from '../../hooks/useChangeUserEmail';
+import useUserStore from '../../stores/useUserStore';
+import { useSnackbarStore } from '../../stores/snackbarStore';
+import SnackbarProvider from '../SnackbarProvider';
+
+function ChangeUserEmail() {
+  const { isSuccess, isError, error, isPending, mutate } = useChangeUserEmail();
+  const { openSnackbar } = useSnackbarStore();
+  const setUser = useUserStore((s) => s.setUser);
+
+  const [emailForm, setEmailForm] = useState<ChangeEmailData>({ newEmail: '' });
+
+  const { data: userResponse, isLoading } = useQuery({
+    queryKey: ['me'],
+    queryFn: fetchUser,
+    retry: false,
+  });
+  const user = userResponse?.data;
+
+  useEffect(() => {
+    if (user) {
+      setEmailForm({
+        newEmail: user.email,
+      });
+    }
+  }, [user]);
+
+  useEffect(() => {
+    if (isSuccess) {
+      openSnackbar('Email successfully changed!', 'success');
+    } else if (isError) {
+      const errMsg =
+        (error as AxiosError<ServerError>)?.response?.data.message ||
+        'Email change error';
+      openSnackbar(errMsg, 'error');
+    }
+  }, [isSuccess, isError, error, openSnackbar]);
+
+  if (isLoading) return <p>Loading...</p>;
+
+  const handleChange =
+    <F extends object>(setter: React.Dispatch<React.SetStateAction<F>>) =>
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      const { name, value } = e.target;
+      setter((prev: F) => {
+        const updated = { ...prev, [name]: value } as F;
+        if (name === 'newEmail') {
+          setUser({
+            email: (updated as ChangeEmailData).newEmail,
+            isLoggedIn: true,
+            firstName: '',
+            lastName: '',
+            isEmailConfirmed: false,
+          });
+        }
+        return updated;
+      });
+    };
```

Comment:
to jest dalej zbyt skomplikowane. nie trzeba robic z tego funkcji generycznej. w `onChange` trzeba zrobic `onChange={() => handleChange()}`. tutaj chcemy tylko zapisac input uzytkownika w jakiejs zmiennej zeby potem moc to wyslac do backendu.
```ts
const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
  emailFrom.newEmail = e.currentTarget.value;
}
```